app/services/course/assessment/submission/auto_grading_service.rb

```diff
@@ -38,6 +50,13 @@ def grade_answers(submission)
   def grade_answer(answer)
     raise ArgumentError if answer.changed?
     answer.auto_grade!
+    # Catch errors if answer is in attempting state, caused by a race condition where
+    # a new attempting answer is created while the submission is finalised, but before the
+    # autograding job is executed.
+  rescue IllegalStateError
```

Comment:
https://github.com/Coursemology/coursemology2/pull/1934 allows graded and evaluated answers to be evaluated again.

For consistency, would it be better to allow attempting answers to be evaluated too? Then there's no need to specially catch an `IllegalStateError`.

I think it'll be clearer to explicitly allow it in the Answer model's state transition definition. That'll also keep the state transition logic in 1 place and prevent unexpected surprises. (The next person might use this elsewhere wonder why grading `attempting` answers doesn't throw an exception like the model definition suggests.)